src/best_offset_prefetcher.cpp

```diff
@@ -0,0 +1,269 @@
+ * the zsim paper ("ZSim: Fast and Accurate Microarchitectural Simulation of
+ * Thousand-Core Systems", Sanchez and Kozyrakis, ISCA-40, June 2013) as the
+ * source of the simulator in any publications that use this software, and that
+ * you send us a citation of your work.
+ *
+ * zsim is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "best_offset_prefetcher.h"
+#define L3_ACCESS_TIME 49
+//#define L3_ACCESS_TIME 27
+//#define L3_ACCESS_TIME 30
+
+// helper to reset offset scores to 0
+void BestOffsetPrefetcher::reset_offsets(){
+    for (auto& i : offset_scores){
+        i.second = 0; 
+    } 
+}
+void BestOffsetPrefetcher::print_scores(){
+    std::cout << std::endl;
+    std::cout << std::endl;
+    for (auto& i : offset_scores){
+        if (i.second != 0)
+            std::cout << "Offset: " << i.first << " Score: " << i.second << std::endl;
+    } 
+}
+
+void BestOffsetPrefetcher::reset_prefetcher(){
+    // return all offsets to 0
+    reset_offsets();
+    // set current round back to 0
+    this->current_round = 0;
+    // set the test offset back to the end index (for timely prefetches)
+    this->test_offset_index = 0;
+}
+
+uint64_t BestOffsetPrefetcher::find_max_score(){
+    std::pair<uint64_t,uint64_t>  max_score = std::pair<uint64_t,uint64_t>(1, BAD_SCORE); 
+    for (auto s : offset_scores){
+        if (s.second >= max_score.second){
+            // set the offset to the offset value
+            max_score.first = s.first;
+            // set the second to the score
+            max_score.second = s.second; 
+        } 
+    }
+    return max_score.first; 
+}
+
+void BestOffsetPrefetcher::move_test_offset_ptr(){
+    // if the current offset is the total minus 1
+    if (this->test_offset_index == (NUM_OFFSETS - 1) ){
+        // reset the pointer and set the round up
+        this->test_offset_index = 0;
+        this->current_round++;
+    // if max out the nuber of rounds, 
+        // reset the prefetcher
+         if (this->current_round == ( RND_MAX + 1 ) ){
+            this->current_offset = find_max_score(); 
+            reset_prefetcher();
+        }
+    }
+     else{
+        // move address in test_offset_point along
+        this->test_offset_index++;
+    }
+}
+
+
+void BestOffsetPrefetcher::learn(uint64_t addr, uint64_t cycle){
+    // offset to currently test
+    uint64_t test_offset = (this->offset_scores[this->test_offset_index]).first;
+    // if we're going off the page, dont' get it
+    uint64_t base_addr = addr - test_offset;
+    // if we're going off the page, dont' get it
+    if ( ((addr - test_offset) < 0 )
```

Comment:
Heiner, I'm not sure what you mean by the next page. If the current access address minus the test offset puts it on another page, then the offset cannot be considered as generating a valid prefetch because the only way that it would hit a previous access would be random chance.